### PR TITLE
Game backend

### LINKED ISSUE #6: BE: Match Controller
Route for creating a new match CREATE /match. When user creates a match, return the match id to the frontend. It will be user so that the player can invite another to join the same match

### LINKED ISSUE #8: BE: Route for joining a match
Route for joining a match. Given a match id, add player to match and return their initial state

IF you're given a match route, for example /match/1234 , I'm joining match id 1234 and when I visit this link, I should be able in the waiting room waiting for other players on this same link

### LINKED ISSUE #19: BE: Game Engine V2
- to_json method : include all the necessary information relayed to the frontend
- assign method: assign one user to one role so that it can be relayed to the frontend

### PR DESCRIPTION
**Pull request by mpmpx**
Added routes and controller for game. It now supports:
  - create a game
  - join in a game
  - assign roles
  - start the game
  - make next move
  - get current game state

Adjusted some methods in game engine according to needs from game handlers.
More adjustment may be made .

### CODE REVIEW

**Comment by mpmpx** [comment]
Added end-turn and restart ,methods, redesigned code structure in back end controller.

File: server/controllers/game.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+const globalState = {};
+
+const validateGameId = (req, res, next) => {
+  const id = req.params.id;
+  if (globalState[id] === undefined)
+    return res.status(404).json({error: `game id ${id} not found`});
+
+  res.locals.game = globalState[id].gameEngine;
+  next();
+}
```
Comment (sina-jamshidi) [inline server/controllers/game.js:12]: nice!

File: server/controllers/game.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  else if (word !== undefined) {
+    res.locals.method = res.locals.game.guesserNextMove;
+    res.locals.params = [player, word];
+  }
+  else {
+    throw new Error('Next move is not provided.');
+  }
+
+  next();
+}
```
Comment (sina-jamshidi) [inline server/controllers/game.js:81]: looks good

File: server/routes/game.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+router.get('/:id/ping', game.ping);
+
+// Store method and parameters in res.locals and pass to last execution
+// middleware.
+router.put('/:id/join', game.join, game.execute);
+router.put('/:id/assign', game.assign, game.execute);
+router.put('/:id/start', game.start, game.execute);
+router.put('/:id/next-move', game.nextMove, game.execute);
+router.put('/:id/end-turn', game.endTurn, game.execute);
+router.put('/:id/restart', game.restart, game.execute);
```
Comment (sina-jamshidi) [inline server/routes/game.js:20]: This is a good infrastructure

**Review by sina-jamshidi — APPROVED** [review]

**Pull request merged**